### PR TITLE
#6664: Catalog service selector for dashboards 

### PR DESCRIPTION
**Pull request by MuhweziDeo**
## Description

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**
Adds Catalog Service capabilities to Dashboard
#6664 

**What is the new behavior?**
The feature is not currently Implemented. 

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/plugins/widgetbuilder/enhancers/connection/catalogEditorConnect.js
Comment (offtherailz) [inline web/client/plugins/widgetbuilder/enhancers/connection/catalogEditorConnect.js:28]: We should not use the catalog state. 
We should instead use internal state of react as much as possible, and the state of the dashboard,you should store only if some change has been really applied:
 - edit and save an existing service
 - create a new service
 - delete an extisting service

So `onChangeUrl`, `onChangeTitle`, `onChangeType`, `onChangeServiceFormat`, `onChangeMetadataTemplate`, `onToggleAdvancedSettings`, `onChangeServiceProperty`, `onToggleTemplate`, `onToggleThumbnail`, `onChangeCatalogMode` should be all internal to react component.

`onAddService` (that is triggered also for update existing services and that should validate the service in react-component) and `onDeleteService` should instead be bind to the dashboard state, to a new action that UPDATE_SERVICES.
In this cases you should save in the state of the dashboard the new list of services (Copying the existing one, with local changes), and use it instead of the configured one. This list is finally saved in dashboard resource

File: web/client/selectors/catalog.js
Comment (offtherailz) [inline web/client/selectors/catalog.js:15]: nothing should stay in catalog (plugjn) state

**Review by offtherailz — CHANGES_REQUESTED** [review]
Looks good. Seems to work very well :+1: 

**bugs:**
- [ ] When select WMTS layer from the add layer of the map, it gives me the tooltip that is valid only for charts.
wrong_tooltip
- [x] `selectedService` should be only the ID of the service (key of the object). Not the entire object. If not present, select the first you find automatically.
- [x] If the configurations are not present, the plugin doesn't work. In order to make this update automatic, if `pluginCfg.services` `pluginCfg.selectedServices` are not present, use the old `catalog` configuration (as unique element of the array, and so selected). If no configuration is present, show the empty view, without crashing. 

**Some changes:** 
- [ ] Please document the new configuration options in JSDoc of DashboardEditor plugin
    - https://github.com/geosolutions-it/MapStore2/blob/master/web/client/plugins/DashboardEditor.jsx#L89
- [x] The epic is not necessary, because there is no validity check yet. Please update the data directly from the original action in the reducer, removing the unnecessary actions ( 1 action of trigger, loading spinner and notification are not necessary, until you do a validity check).
- [x] Document the actions you added.

**some notes** 
I'm going to merge this. Probably you will see some conflict after this merge, please solve them.
https://github.com/geosolutions-it/MapStore2/pull/6619

**Comment by MuhweziDeo** [comment]
> Looks good. Seems to work very well 👍
> 
> **bugs:**
> 
> * [ ]  When select WMTS layer from the add layer of the map, it gives me the tooltip that is valid only for charts.
>   wrong_tooltip
> * [ ]  `selectedService` should be only the ID of the service (key of the object). Not the entire object. If not present, select the first you find automatically.
> * [ ]  If the configurations are not present, the plugin doesn't work. In order to make this update automatic, if `pluginCfg.services` `pluginCfg.selectedServices` are not present, use the old `catalog` configuration (as unique element of the array, and so selected). If no configuration is present, show the empty view, without crashing.
> 
> **Some changes:**
> 
> * [ ]  Please document the new configuration options in JSDoc of DashboardEditor plugin
>   
>   * https://github.com/geosolutions-it/MapStore2/blob/master/web/client/plugins/DashboardEditor.jsx#L89
> * [ ]  The epic is not necessary, because there is no validity check yet. Please update the data directly from the original action in the reducer, removing the unnecessary actions ( 1 action of trigger, loading spinner and notification are not necessary, until you do a validity check).
> * [ ]  Document the actions you added.
> 
> **some notes**
> I'm going to merge this. Probably you will see some conflict after this merge, please solve them.
> #6619

thanks for the feedback

**Comment by MuhweziDeo** [comment]
Hi @offtherailz not sure where this issue is comming from but I see there is an error in the console `When select WMTS layer from the add layer of the map, it gives me the tooltip that is valid only for charts`. When making the request from WMTS and this sets the `canProceed to False

**Comment by offtherailz** [comment]
@MuhweziDeo please do not force-push pull requests where possible.
It makes review process hard, because the reviewer have to re-start from the beginning, not being able to see what are changes from last review

File: web/client/localConfig.json
Comment (offtherailz) [inline web/client/localConfig.json:42]:
```suggestion
      },
```

File: web/client/plugins/DashboardEditor.jsx
Comment (offtherailz) [inline web/client/plugins/DashboardEditor.jsx:92]:
```suggestion
* @prop {object} cfg.services Object with the catalogs available to select layers for maps, charts and tables. The format is the same of the `Catalog` plugin.
```

**Comment by MuhweziDeo** [comment]
> @MuhweziDeo please do not force-push pull requests where possible.
> It makes review process hard, because the reviewer have to re-start from the beginning, not being able to see what are changes from last review

Hey, @offtherailz Sorry about that. Will make sure not to repeat it

File: web/client/plugins/DashboardEditor.jsx
Comment (offtherailz) [inline web/client/plugins/DashboardEditor.jsx:92]:
```suggestion
* @prop {object} cfg.catalog **Deprecated** in favor of `cfg.services`. Can contain a catalog configuration
* @prop {object} cfg.services Object with the catalogs available to select layers for maps, charts and tables. The format is the same of the `Catalog` plugin.
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
- [x] Search on service selection option in advanced is persent, but is not needed. Also `autoload` option is not needed in configuration
- [x] You can remove the old `catalog` setting in local configuration, in favor of the new one.
- [ ] TMS and WFS catalog types do not seem to work (tried adding from config or from UI)
- [x] When I add a layer to a map (tested with empty map) I have this error:
```
manageLayers.js:67 Uncaught TypeError: Cannot read property 'localizedLayerStyles' of undefined
    at eval (manageLayers.js:67)
    at eval (Recompose.js:356)
    at eval (MapBuilder.jsx:96)
    at eval (Recompose.js:356)
    at onClick (MapLayerSelector.jsx:90)
    at onClick (buttonWithDisabled.jsx:37)
```
- [x] When save a new catalog from map widget builder, I'm redirect to TOC instead of continue selecting layer.

- [ ] I suggest also to handle `mode` and isNew inside the component state, to avoid possible state inconsistencies (the global state is not reset, while the component state is. This helps to restart from the beginning).
For instance create a new Map widget --> empty map --> add layer --> add catalog --> then click on home button to get back to the home page (the state of "editing" stays there). when you get back to the dashboard, adding the layer and so on, you will see again the catalog in edit mode.

File: web/client/plugins/widgetbuilder/CatalogServiceEditor.jsx
Comment (offtherailz) [inline web/client/plugins/widgetbuilder/CatalogServiceEditor.jsx:30]:
```suggestion
        if (!service.type !== 'tms' && !(!service.title || !service.url)) { // TODO validation should be internal to catalog editor, in common.
```

Comment (MuhweziDeo) [reply web/client/plugins/widgetbuilder/CatalogServiceEditor.jsx:30]: Okay not sure what this means @offtherailz

File: web/client/components/catalog/CompactCatalog.jsx
Comment (offtherailz) [inline web/client/components/catalog/CompactCatalog.jsx:47]:
```suggestion
const getIdentifier = (r) => r.identifier ? r.identifier :  r.provider ? r.provider + (r.variant ?? "") : (r.title || r.url);
```

File: web/client/plugins/widgetbuilder/commons.js
Comment (offtherailz) [inline web/client/plugins/widgetbuilder/commons.js:49]: can these be placed in catalogEditorSelector too?

**Review by offtherailz — CHANGES_REQUESTED** [review]
- [ ] See comments inline about variables that can be moved in the proper enhancer
- [ ] trying to edit an extisting wfs service causes crash (localizedProp = true), using this config and trthing to edit the service `GeoSolutions GeoServer WFS`
```
"selectedService": "gs_stable_csw",
            "services": {
              "gs_stable_csw": {
                "url": "https://gs-stable.geo-solutions.it/geoserver/csw",
                "type": "csw",
                "title": "GeoSolutions GeoServer CSW",
                "autoload": true
              },
              "gs_stable_wms": {
                "url": "https://gs-stable.geo-solutions.it/geoserver/wms",
                "type": "wms",
                "title": "GeoSolutions GeoServer WMS",
                "autoload": false
              },
              "gs_stable_wfs": {
                "url": "https://gs-stable.geo-solutions.it/geoserver/wfs",
                "type": "wfs",
                "title": "GeoSolutions GeoServer WFS",
                "autoload": false
              },
              "gs_stable_wmts": {
                "url": "https://gs-stable.geo-solutions.it/geoserver/gwc/service/wmts",
                "type": "wmts",
                "title": "GeoSolutions GeoServer WMTS",
                "autoload": false
              }
```
- [ ] layers sometimes can not be added to the map because "they are not raster". That limitation should not be present in the map use case.

**Review by offtherailz — CHANGES_REQUESTED** [review]
I need to be able to test do finalize the review, but I encountered this bug on the very first attempt
- [x] As well as I open the catalog, select the first layer of csw, it crashes. it looks that dashboardServices is undefined, maybe you have to populate the dashboardServices with defaults somewhere (in the selector or in the main comoment)

Please test it extensively to check that there is no error before submitting the review (every error that can happen in the UI should be managed) so we can avoid slows down, thank you

**Comment by MuhweziDeo** [comment]
> I need to be able to test do finalize the review, but I encountered this bug on the very first attempt
> 
> * [ ]  As well as I open the catalog, select the first layer of csw, it crashes. it looks that dashboardServices is undefined, maybe you have to populate the dashboardServices with defaults somewhere (in the selector or in the main comoment)
> 
> Please test it extensively to check that there is no error before submitting the review (every error that can happen in the UI should be managed) so we can avoid slows down, thank you

Okay

File: web/client/plugins/widgetbuilder/enhancers/layerSelector.js
Comment (offtherailz) [inline web/client/plugins/widgetbuilder/enhancers/layerSelector.js:18]: I fixed a little bit this part. The only part is missing and you have to add is the remove of autkey param. See epics/catalog.js. It uses authkeyParamNameSelector to get and call recordToLayer and remove authkey from params. You have to do the same, and add this to your recordToLayer Call, as third parameter

```suggestion

// layers like tms or wmts don't need the recordToLayer conversion
const toLayer = (r, service) => ["tms", "wfs"].includes(service.type) // for tms and wfs the layer is ready
    ? r : recordToLayer(
        r, service.type === "wmts" // the type wms is default (for csw and wms), wmts have to be passed. // TODO: improve and centralize more
            ? "wmts"
            : undefined
    ); //

// checks for tms wmts in order to addSearch() to skip addSearch
const addSearchObservable = (selected, service) => ["tms", "wmts"].includes(service.type) ? Rx.Observable.of(toLayer(selected, service)) : addSearch(toLayer(selected, service));
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
Testing I had some problems with wfs catalog, that was crashing, and WMTS layers , that was not visible. 
I fixed them in order to work (even if the record conversion need some simplification, but this is not the scope of this work).

For this reason I applied some fixes and it should work (when merge my inline suggestion) . The only thing missing is the authkeyParamName removal (see the comment inline).

**Review by offtherailz — CHANGES_REQUESTED** [review]
Is this PR still under development @MuhweziDeo or should I review ? 
I tried it and I found some issues I have notify so you can fix: 
- Add layers from TMS services in map widgets not working properly:
- TMS of type 1.0.0 selection not working (it selects more than one) I fixed it a little applying this fix. I think it can work, even if we can do it better.
```javascript
/*
 * assigns an identifier to a record. The ID is required for local selection.
 * TODO: improve identifier generation.
 */
const getIdentifier = (r) =>
    r.identifier ? r.identifier
        :  r.provider
            ? r.provider + (r.variant ?? "") // existing tileprovider
            : (r.tileMapUrl // TMS 1.0.0
        || r.url); //  default
```
- I created a dashboard (with custom services) than reladed, tried to add a new map widget. As well as I try to add one layer to the map, it crashes with the following error.
Peek 03-05-2021 14-31

```
StandardRouter.jsx:88 TypeError: Cannot read property 'type' of undefined
    at toLayer (webpack://mapstore2/./web/client/plugins/widgetbuilder/enhancers/layerSelector.js?:23)
    at SwitchMapSubscriber.eval [as project] (webpack://mapstore2/./web/client/plugins/widgetbuilder/enhancers/layerSelector.js?:47)
    at SwitchMapSubscriber._next (webpack://mapstore2/./node_modules/rxjs/operator/switchMap.js?:88)
    at SwitchMapSubscriber.Subscriber.next (webpack://mapstore2/./node_modules/rxjs/Subscriber.js?:89)
    at FilterSubscriber._next (webpack://mapstore2/./node_modules/rxjs/operator/filter.js?:88)
    at FilterSubscriber.Subscriber.next (webpack://mapstore2/./node_modules/rxjs/Subscriber.js?:89)
    at DistinctUntilChangedSubscriber._next (webpack://mapstore2/./node_modules/rxjs/operator/distinctUntilChanged.js?:103)
    at DistinctUntilChangedSubscriber.Subscriber.next (webpack://mapstore2/./node_modules/rxjs/Subscriber.js?:89)
    at Array.eval (webpack://mapstore2/./node_modules/recompose/es/Recompose.js?:988)
    at Object.emit (webpack://mapstore2/./node_modules/change-emitter/lib/index.js?:43)
    at ComponentFromStream.componentWillReceiveProps (webpack://mapstore2/./node_modules/recompose/es/Recompose.js?:1020)
    at callComponentWillReceiveProps (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:14329)
    at updateClassInstance (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:14546)
    at updateClassComponent (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:18357)
    at beginWork$1 (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:20108)
    at beginWork$$1 (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:25708)
``` 
- Again, please do not  force push, or the previous reviews will be thrown away and I have to start the review from the beginning.

**Comment by MuhweziDeo** [comment]
> Is this PR still under development @MuhweziDeo or should I review ?
> I tried it and I found some issues I have notify so you can fix:
> 
> * Add layers from TMS services in map widgets not working properly:
> * TMS of type 1.0.0 selection not working (it selects more than one) I fixed it a little applying this fix. I think it can work, even if we can do it better.
> 
> ```js
> /*
>  * assigns an identifier to a record. The ID is required for local selection.
>  * TODO: improve identifier generation.
>  */
> const getIdentifier = (r) =>
>     r.identifier ? r.identifier
>         :  r.provider
>             ? r.provider + (r.variant ?? "") // existing tileprovider
>             : (r.tileMapUrl // TMS 1.0.0
>         || r.url); //  default
> ```
> 
> * I created a dashboard (with custom services) than reladed, tried to add a new map widget. As well as I try to add one layer to the map, it crashes with the following error.
>   Peek 03-05-2021 14-31
> 
> ```
> StandardRouter.jsx:88 TypeError: Cannot read property 'type' of undefined
>     at toLayer (webpack://mapstore2/./web/client/plugins/widgetbuilder/enhancers/layerSelector.js?:23)
>     at SwitchMapSubscriber.eval [as project] (webpack://mapstore2/./web/client/plugins/widgetbuilder/enhancers/layerSelector.js?:47)
>     at SwitchMapSubscriber._next (webpack://mapstore2/./node_modules/rxjs/operator/switchMap.js?:88)
>     at SwitchMapSubscriber.Subscriber.next (webpack://mapstore2/./node_modules/rxjs/Subscriber.js?:89)
>     at FilterSubscriber._next (webpack://mapstore2/./node_modules/rxjs/operator/filter.js?:88)
>     at FilterSubscriber.Subscriber.next (webpack://mapstore2/./node_modules/rxjs/Subscriber.js?:89)
>     at DistinctUntilChangedSubscriber._next (webpack://mapstore2/./node_modules/rxjs/operator/distinctUntilChanged.js?:103)
>     at DistinctUntilChangedSubscriber.Subscriber.next (webpack://mapstore2/./node_modules/rxjs/Subscriber.js?:89)
>     at Array.eval (webpack://mapstore2/./node_modules/recompose/es/Recompose.js?:988)
>     at Object.emit (webpack://mapstore2/./node_modules/change-emitter/lib/index.js?:43)
>     at ComponentFromStream.componentWillReceiveProps (webpack://mapstore2/./node_modules/recompose/es/Recompose.js?:1020)
>     at callComponentWillReceiveProps (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:14329)
>     at updateClassInstance (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:14546)
>     at updateClassComponent (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:18357)
>     at beginWork$1 (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:20108)
>     at beginWork$$1 (webpack://mapstore2/./node_modules/react-dom/cjs/react-dom.development.js?:25708)
> ```
> 
> * Again, please do not  force push, or the previous reviews will be thrown away and I have to start the review from the beginning.

When I resolve merge conflicts I normally do a rebase and with that I have to force push donot know any other way

**Review by offtherailz — CHANGES_REQUESTED** [review]
TMS layers still not working. Please try to use all types of catalog and all types of TMS and verify that every of that works properly. You can use the normal catalog in maps as example. 
For charts/counters/tables: 
- [ ] WMS (some of layers should give the advice of not compatible)
- [ ] WFS
- [ ] CSW

The other catalogs types should work but should always give the advice or not compatible. (we can improve in the future to hide them in charts/tables)

For map make you sure that you can add layers (and the layers are visible on the map) from
- [ ] WMS
- [ ] WMTS
- [ ] CSW
- [ ] WFS
- [ ] TMS
   - [ ] 1.0.0 (e.g .https://demo.geo-solutions.it/geoserver/gwc/service/tms/1.0.0 ). Add layer of projection 900913.
   - [ ] custom (use this for instance as custom URL https://dev.{s}.tile.openstreetmap.fr/cyclosm/{z}/{x}/{y}.png)
   - [ ] Stamen and OpenSeaMap

Here the problem I noticed
Peek 04-05-2021 10-33

**Review by offtherailz — APPROVED** [review]
:+1:

**Pull request merged**
